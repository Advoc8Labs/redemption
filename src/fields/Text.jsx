import React from 'react';

/**
 * Disambiguation: this is NOT a text "field" that receives input.
 * It just models a block of text in a form.
 */
function Text({ text }) {
    const { text_formatted: formattedText } = text;

    // TODO: Make this in React
    const nodes = formattedText.ops
        .map((op) => {
            const attrs = op.attributes;

            const node = document.createElement(attrs?.link ? 'a' : 'span');
            node.innerHTML = op.insert.replace(/\n/g, '<br>');

            if (attrs) {
                if (attrs.size) {
                    node.style.fontSize = attrs.size;
                }

                if (attrs.font) {
                    node.style.fontFamily = attrs.fontFull.replace(/"/g, "'");
                }

                if (attrs.color) {
                    node.style.color = attrs.color;
                }

                if (attrs.weight) {
                    node.style.fontWeight = attrs.weight;
                }

                if (attrs.italic) {
                    node.style.fontStyle = 'italic';
                }

                if (attrs.link) {
                    node.href = attrs.link;
                }

                const lines = [];
                if (attrs.strike) {
                    lines.push('line-through');
                }

                if (attrs.underline) {
                    lines.push('underline');
                }

                if (lines.length > 0) {
                    node.style.textDecoration = lines.join(' ');
                }
            }

            return node.outerHTML;
        })
        .join('');

    return <div dangerouslySetInnerHTML={{ __html: nodes }} />;
}

export default Text;
