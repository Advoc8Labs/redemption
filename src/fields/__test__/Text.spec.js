import React from 'react';
import { create, act } from 'react-test-renderer';
import Text from '../Text';

describe('Text', () => {
    it('renders an empty text block', async () => {
        // Arrange
        const props = {
            text: {
                text_formatted: {
                    ops: [{ insert: '\n' }]
                }
            }
        };

        // Act
        let text;
        act(() => {
            text = create(<Text {...props} />);
        });
        const tree = text.toJSON();

        // Assert
        expect(tree).toMatchSnapshot();
    });

    it('renders a plaintext text block', async () => {
        // Arrange
        const props = {
            text: {
                text_formatted: {
                    ops: [{ insert: 'Hello World!' }]
                }
            }
        };

        // Act
        let text;
        act(() => {
            text = create(<Text {...props} />);
        });
        const tree = text.toJSON();

        // Assert
        expect(tree).toMatchSnapshot();
    });

    it('renders a text block with custom styles', async () => {
        // Arrange
        const props = {
            text: {
                text_formatted: {
                    ops: [
                        {
                            insert: 'Hello World!',
                            attributes: {
                                weight: 700,
                                color: '#000000',
                                italic: true,
                                size: '48px',
                                font: 'sans',
                                fontFull: 'sans-serif'
                            }
                        }
                    ]
                }
            }
        };

        // Act
        let text;
        act(() => {
            text = create(<Text {...props} />);
        });
        const tree = text.toJSON();

        // Assert
        expect(tree).toMatchSnapshot();
    });
});
