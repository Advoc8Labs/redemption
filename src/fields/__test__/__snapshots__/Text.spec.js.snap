// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text renders a plaintext text block 1`] = `
<div
  dangerouslySetInnerHTML={
    Object {
      "__html": "<span>Hello World!</span>",
    }
  }
/>
`;

exports[`Text renders a text block with custom styles 1`] = `
<div
  dangerouslySetInnerHTML={
    Object {
      "__html": "<span style=\\"font-size: 48px; font-family: sans-serif; color: rgb(0, 0, 0); font-weight: 700; font-style: italic;\\">Hello World!</span>",
    }
  }
/>
`;

exports[`Text renders an empty text block 1`] = `
<div
  dangerouslySetInnerHTML={
    Object {
      "__html": "<span><br></span>",
    }
  }
/>
`;
